File: csvrow.lua
#!/usr/bin/env lua

--[[ csvrow.lua, v1.0

Extracts a specified record from a CSV file, local or remote, and
prints it in an INI-like format.

Usage: csvrow.lua <key> <csv-input>

Arguments:
  <key>         Specifies a value matching the first field in the target row.
  <csv-input>   Specifies a URL or a file corresponding to CSV data, which
                may contain data or a URL to the data.

------------------------------------------------------------------------------]]

local sep = "="
local exit_status = 0

-- Function to check if a string is a URL
function is_url(str)
    local pattern = "^https?://[%w-_%.%?%.:/%+=&]+$"
    return string.match(str, pattern) ~= nil
end

-- Function to retrieve CSV content from a local or remote source
function get_csv(input)
    local command
    if is_url(input) then
        command = "rclone copyurl --stdout " .. input
    else
        command = "rclone cat " .. input
    end

    local handle = io.popen(command)
    local csv_content = handle:read("*a")
    handle:close()
    
    -- Normalize line endings to Unix style
    csv_content = csv_content:gsub("\r\n", "\n"):gsub("\r", "\n")

    -- Trim any leading/trailing whitespace from the content
    csv_content = csv_content:match("^%s*(.-)%s*$")
    
    if is_url(csv_content) then
        return get_csv(csv_content)
    else
        return csv_content
    end
end

-- Function to parse CSV string into a table of lines
function csv_to_table(csv_string)
    local result = {}
    for line in csv_string:gmatch("[^\r\n]+") do
        table.insert(result, line)
    end
    return result
end

-- Function to convert a CSV record to INI-like format
function csv_record_to_ini(csv_lines, match)
    local headers = {}
    local values = {}
    for i, line in ipairs(csv_lines) do
        local fields = {}
        for field in line:gmatch("([^,]+)") do
            table.insert(fields, field)
        end
        if i == 1 then
            headers = fields
        elseif fields[1] == match then
            values = fields
            break
        end
    end

    if #values == 0 then
        return nil
    end

    local ini_content = {}
    for i = 1, #headers do
        if string.upper(headers[i]) == headers[i] then
            local line = string.format("%s%s%s", headers[i], sep, values[i])
            table.insert(ini_content, line)
        end
    end

    if #ini_content == 0 then
        return nil -- For edge case when all headers not in preferred allcaps
    else
        return table.concat(ini_content, "\n")
    end
end

-- Function to print usage
function usage()
    print("csvrow.lua 1.0, prints specified CSV row data alongside headers " ..
          "in INI-like format.")
    print("Usage: csvrow.lua <key> <csv-input>\n")
    print("Where:")
    print("  <key>         Specifies a value matching the first field " ..
                           "in the target row.")
    print("  <csv-input>   Specifies a URL or a file corresponding to " .. 
                           "CSV data, which")
    print("                may contain data or a URL to the data.\n")
end

if #arg < 2 then
    usage()
    os.exit(1)
end

local record_to_match = arg[1]
local input = arg[2]

local csv_content = get_csv(input)
local csv_lines = csv_to_table(csv_content)
local ini_output = csv_record_to_ini(csv_lines, record_to_match)

if ini_output then
    print(ini_output)
else
    -- Do nothing (no ouput)
    exit_status = 1
end
os.exit(exit_status)

File: csvrow.go
package main

import (
  "os"
  "flag"
  "fmt"
  "os/exec"
  "regexp"
  "strings"
  "path/filepath"
)

var usage = `[progName] - extracts key-value data from CSVs

Usage: [progName] [options] <row_selector> <file_locator>

Arguments:
  <row_selector>
         Value from first field in target row
  <file_locator>
         Local or remote path to file containing CSV data
         or URL pointing to CSV data

Options:
  -s string
        Separator between key and value (default "=")
  -v    Enable verbose output

Note:
  The <file_locator> argument supports rclone remotes as
  paths to CSV data. File may contain URL to CSV data,
  however, cannot point to another nested URL.
`

// ------------------------------------------------------------------------

func isURL(input string) bool {
  return strings.HasPrefix(input, "http://") ||
    strings.HasPrefix(input, "https://")
}

// ------------------------------------------------------------------------

func getFileContent(fileLocator string) (string, error) {
    // Define a closure for the recursive logic to prevent infinite loop
    var fetchContent func(fileLocator string, depth int) (string, error)
    
    fetchContent = func(fileLocator string, depth int) (string, error) {
        if depth > 2 { // Prevent infinite recursion
            return "", fmt.Errorf("recursion limit reached")
        }

        // Determine the appropriate rclone command
        var cmd *exec.Cmd
        if isURL(fileLocator) {
            cmd = exec.Command("rclone", "copyurl", fileLocator, "--stdout")
        } else {
            cmd = exec.Command("rclone", "cat", fileLocator)
        }

        data, err := cmd.Output()

        // Handle errors after fetching data
        if err != nil {
            return "", fmt.Errorf("unable to fetch content: %w", err)
        }

        // Normalize line endings and trim whitespace
        content := strings.ReplaceAll(string(data), "\r\n", "\n")
        content = strings.TrimSpace(content)

        if isURL(content) {
            return fetchContent(content, depth + 1)
        } else {
            return content, nil
        }
    }
    
    // Call the closure with initial depth 0
    return fetchContent(fileLocator, 0)
}

// ------------------------------------------------------------------------

func getRecordByKey(lines []string, rowSelector string) (string, bool) {
  pattern := "^" + rowSelector + "," // Build pattern with rowSelector
  regex := regexp.MustCompile(pattern)
  for _, line := range lines {
    if regex.MatchString(line) {
      return line, true
    }
  }
  return "", false
}

// ------------------------------------------------------------------------

func main() {
  var separator string
  var verbose bool

  // Process commandline flags and args with flag package
  // ----------------------------------------------------------------------
  flag.StringVar(&separator, "s", "=", "") // flag desc unused
  flag.BoolVar(&verbose, "v", false, "")   // flag deec unused
  flag.Usage = func() {
    progName := filepath.Base(os.Args[0])
    usage = strings.ReplaceAll(usage, "[progName]", progName)
    fmt.Fprintf(os.Stderr, "%s\n", usage)
  }
  flag.Parse()

  args := flag.Args()

  if len(args) < 2 {
    flag.Usage()
    return
  }

  rowSelector := args[0]
  fileLocator := args[1]

  // ------------------------------------------------------------------------
  // Fetch CSV content
  csvContent, err := getFileContent(fileLocator)
  if err != nil {
    fmt.Fprintf(os.Stderr, "Error: %s\n", err)
    return
  }

  if verbose {
    fmt.Fprintf(os.Stderr, "Fetched CSV content:\n")
    fmt.Fprintf(os.Stderr, "%s\n\n", csvContent)
  }

  // ------------------------------------------------------------------------
  // Process CSV content
  lines := strings.Split(csvContent, "\n")

  if len(lines) < 2 {
    if verbose {
      fmt.Fprintf(os.Stderr, "Warning: CSV contains insuffient content.\n")
    }
    return
  }

  headerLine := lines[0]
  recordLine, found := getRecordByKey(lines[1:], rowSelector)

  if found {
    headers := strings.Split(headerLine, ",")
    values := strings.Split(recordLine, ",")

    for i, header := range headers {
      if strings.ToUpper(header) == header { // Consider only allcaps headers
        if i < len(values) { // Ensure index is within values bounds
          fmt.Printf("%s%s%s\n", header, separator, values[i])
        } else {
          fmt.Printf("%s%s\n", header, separator)
          if verbose {
            fmt.Fprintf(os.Stderr, "Warning: %s has no corresponding value.\n", header)
          }
        }
      }
    }

  } else if verbose {
    fmt.Fprintf(os.Stderr, "Waning: Pattern not found.\n")
  }
}
